biz.aQute.remote/src/aQute/remote/util/AgentSupervisor.java

```diff
@@ -41,16 +42,26 @@
 	}
 
 	protected void connect(Class<Agent> agent, Supervisor supervisor, String host, int port) throws Exception {
-		while (true)
+		connect(agent, supervisor, host, port, -1);
+	}
+
+	protected void connect(Class<Agent> agent, Supervisor supervisor, String host, int port, int timeout)
+			throws Exception {
+		while (true) {
 			try {
 				Socket socket = new Socket(host, port);
 				link = new Link<Supervisor,Agent>(agent, supervisor, socket);
 				this.setAgent(link);
 				link.open();
 				return;
 			} catch (ConnectException e) {
-				Thread.sleep(200);
+				if (timeout == 0) {
+					return;
+				}
+				timeout = Math.max(timeout - connectWait, 0);
```

Comment:
Need to handle `timeout == -1`. We don't want to convert it to 0 and then return on the next iteration.
```java
if (timeout > 0) {
   timeout = Math.max(timeout - connectWait, 0);
}
```